packages/plugin-inline/Inline.ts

```diff
@@ -0,0 +1,106 @@
+
+export interface FormatParams extends RangeParams {
+    format: Format;
+}
+
+export class Inline extends JWPlugin {
+    commands = {
+        toggleFormat: {
+            handler: this.toggleFormat.bind(this),
+        },
+    };
+    commandHooks = {
+        setSelection: this.resetFormatCache.bind(this),
+    };
+    /**
+     * When apply format on a collapsed range, cache the calculation of the
+     * format the following property. This value is reset each time the range
+     * change in a document.
+     */
+    formatCache: Format[] = null;
+
+    //--------------------------------------------------------------------------
+    // Public
+    //--------------------------------------------------------------------------
+
+    /**
+     * Apply the `format` to the selection.
+     *
+     * If the selection is collapsed, set the format on the selection so we know
+     * in the next insert which format should be used.
+     */
+    toggleFormat(params: FormatParams): void {
+        const format = params.format;
+        const range = this.editor.vDocument.selection.range;
+        if (range.isCollapsed()) {
+            if (!this.formatCache) {
+                this.formatCache = this.getCurrentFormats();
+            }
+            const index = this.formatCache.findIndex(f => f.name === format.name);
+            if (index !== -1) {
+                this.formatCache.splice(index, 1);
+            } else {
+                this.formatCache.push(format.shallowDuplicate());
+            }
+        } else {
+            const selectedInlines = range.selectedNodes(InlineNode);
+
+            // If every char in the range has the format `format.name`, remove
+            // the format for all of them.
+            const allHaveFormat = selectedInlines.every(inline => {
+                return !!inline.formats.find(f => f.name === format.name);
+            });
+            if (allHaveFormat) {
+                selectedInlines.forEach(inline => {
+                    const index = inline.formats.findIndex(f => f.name === format.name);
+                    const matchingFormat = inline.formats[index];
+                    if (matchingFormat.attributes && matchingFormat.attributes.size) {
+                        matchingFormat.attributes.forEach((value: string, key: string) => {
+                            if (!inline.attributes) {
+                                inline.attributes = new Map<string, string>();
+                            }
+                            inline.attributes.set(key, value);
+                        });
+                    }
+                    inline.formats.splice(index, 1);
+                });
+                // If there is at least one char in the range without the format
+                // `format.name`, set the format for all nodes.
+            } else {
+                selectedInlines.forEach(inline => {
+                    if (!inline.formats.find(f => f.isSameAs(format))) {
```

Comment:
It's a bit weird because different nodes might have `format` in different places in their `formats` array.
Consider this:
`<b><i>ab</i></b>[c]<b><i>de</i></b>`

Toggle format bold should give (I know we're not handling this yet but bear with me):
`<b><i>ab</i>[c]<i>de</i></b>`

While this:
`<i><b>ab</b></i>[c]<i><b>de</b></i>`
Toggle format bold should give:
`<i><b>ab</b></i><b>c</b><i><b>de</b></i>`

What I mean is that the order in the `formats` array matters, doesn't it ?

Reply:
Out of scope but we should handle order.